File: Clones/zombia/javascript_server_src/Network/Rpcs/KickMember.js
module.exports = KickMember = (socket, server, data) => {
    const playerEntity = server.entities[socket.uid];
    const playerParty = server.parties[playerEntity.partyId];
    const kickedMemberEntity = server.entities[data.uid];

    if (kickedMemberEntity === undefined) return;

    // Developers can't be kicked
    if (kickedMemberEntity.developerMode == true) return;

    // Ensure the kicker is the leader of the party
    if (playerParty.leaderUid !== playerEntity.uid) return;

    // Ensure the kick-receiver is a member of the party
    if (kickedMemberEntity.partyId !== playerParty.id) return;

    // Ensure the player is not kicking themselves
    if (data.uid == playerEntity.uid) return;

    playerParty.removeMember(server, kickedMemberEntity);
    server.createParty(kickedMemberEntity);
}

File: Archive/Clones/zombia/rust_server_src/network/handle_rpc_types/KickMember.rs
use crate::entity_manager::entity_types::Player;
use crate::entity_manager::{entity_types::EntityTypeEnum, manager};
use crate::network::decode_rpc_types::KickMember::KickMemberRpc;
use crate::{PARTIES, create_party};

pub fn handle_rpc(player_entity: &Player, rpc: KickMemberRpc) {
    let player_to_kick = manager::get_entity(rpc.uid);

    let player_to_kick = match player_to_kick {
        Some(EntityTypeEnum::Player(player_entity)) => player_entity,
        _ => return
    };

    if player_entity.party_id != player_to_kick.party_id {
        return;
    }

    PARTIES.with(|parties| {
        let mut parties = parties.borrow_mut();

        let party = parties.get_mut(&player_entity.party_id).unwrap();

        // is the player sending the packet the leader?
        if party.leader_uid != player_entity.generic_entity.uid {
            return;
        }

        // is the player trying to kick themselves?
        if rpc.uid == player_entity.generic_entity.uid {
            return;
        }

        party.remove_member(rpc.uid, false);
    });

    create_party(rpc.uid);
}